File: bench/runners/python/bench.py
#!/usr/bin/env python3
"""FileFS Python (CPython C-extension) bench runner."""

from __future__ import annotations

import json
import sys
import tempfile
import time
from pathlib import Path

import filefs

ITERS = int(sys.argv[1]) if len(sys.argv) > 1 else 40
WARMUP = int(sys.argv[2]) if len(sys.argv) > 2 else 2
PAYLOAD = 4096


def median(samples: list[float]) -> float:
    samples = sorted(samples)
    n = len(samples)
    if n == 0:
        return 0.0
    if n % 2 == 0:
        return (samples[n // 2 - 1] + samples[n // 2]) / 2
    return samples[n // 2]


def time_op(prep, body):
    for _ in range(WARMUP):
        if prep:
            prep()
        body()
    samples = []
    for _ in range(ITERS):
        if prep:
            prep()
        t0 = time.perf_counter_ns()
        body()
        samples.append(float(time.perf_counter_ns() - t0))
    return {"ns_per_op": median(samples), "iters": ITERS}


def main() -> None:
    root = Path(tempfile.mkdtemp(prefix="filefs-bench-py-"))
    image = root / "bench.ffs"
    counter = {"n": 0}

    def uniq(prefix: str) -> str:
        counter["n"] += 1
        return f"{prefix}{counter['n']}"

    payload = bytes(i & 0xFF for i in range(PAYLOAD))
    ops = {}

    def mkfs_once():
        p = root / f"{uniq('mkfs')}.ffs"
        filefs.mkfs(str(p))
        p.unlink(missing_ok=True)
        Path(str(p) + "-j").unlink(missing_ok=True)

    ops["mkfs"] = time_op(None, mkfs_once)

    filefs.mkfs(str(image))
    ffs = filefs.create()
    filefs.mount(ffs, str(image))

    def prep_umount():
        filefs.umount(ffs)

    def mount_umount():
        filefs.mount(ffs, str(image))
        filefs.umount(ffs)

    ops["mount_umount"] = time_op(prep_umount, mount_umount)
    filefs.mount(ffs, str(image))

    ops["mkdir"] = time_op(None, lambda: filefs.mkdir(ffs, uniq("d")))
    filefs.mkdir(ffs, "cwdbench")
    ops["chdir_getcwd"] = time_op(
        None,
        lambda: (filefs.chdir(ffs, "cwdbench"), filefs.getcwd(ffs), filefs.chdir(ffs, "/")),
    )

    def open_close():
        f = filefs.fopen(ffs, f"{uniq('o')}.txt", "w")
        filefs.fclose(ffs, f)

    ops["open_write_close"] = time_op(None, open_close)

    seed = filefs.fopen(ffs, "seed.bin", "w")
    filefs.fwrite(ffs, payload, 1, len(payload), seed)
    filefs.fclose(ffs, seed)

    def write4():
        f = filefs.fopen(ffs, "wbench.bin", "w")
        filefs.fwrite(ffs, payload, 1, len(payload), f)
        filefs.fclose(ffs, f)

    ops["write_4kib"] = time_op(None, write4)

    def read4():
        f = filefs.fopen(ffs, "seed.bin", "r")
        filefs.fread(ffs, 1, PAYLOAD, f)
        filefs.fclose(ffs, f)

    ops["read_4kib"] = time_op(None, read4)

    def seek_ops():
        f = filefs.fopen(ffs, "seed.bin", "r")
        filefs.fseek(ffs, f, 0, filefs.SEEK_END)
        filefs.ftell(ffs, f)
        filefs.rewind(ffs, f)
        filefs.fclose(ffs, f)

    ops["seek_tell_rewind"] = time_op(None, seek_ops)
    ops["copy_file"] = time_op(
        None,
        lambda: (
            filefs.remove(ffs, "copy_dst.bin") if filefs.file_exist(ffs, "copy_dst.bin") else None,
            filefs.copy(ffs, "seed.bin", "copy_dst.bin"),
        ),
    )

    def do_rename():
        src = f"{uniq('r')}.txt"
        dst = f"{uniq('s')}.txt"
        f = filefs.fopen(ffs, src, "w")
        filefs.fclose(ffs, f)
        filefs.rename(ffs, src, dst)
        filefs.remove(ffs, dst)

    ops["rename"] = time_op(None, do_rename)

    def do_remove():
        name = f"{uniq('m')}.txt"
        f = filefs.fopen(ffs, name, "w")
        filefs.fclose(ffs, f)
        filefs.remove(ffs, name)

    ops["remove_file"] = time_op(None, do_remove)

    def readdir():
        handle = filefs.opendir(ffs, "/")
        dir_handle = handle[0] if isinstance(handle, tuple) else handle
        while filefs.readdir(ffs, dir_handle) is not None:
            pass
        filefs.closedir(ffs, dir_handle)

    ops["readdir"] = time_op(None, readdir)
    ops["exists"] = time_op(
        None,
        lambda: (filefs.file_exist(ffs, "seed.bin"), filefs.dir_exist(ffs, "cwdbench")),
    )

    def txn():
        filefs.begin(ffs)
        f = filefs.fopen(ffs, f"{uniq('t')}.txt", "w")
        filefs.fwrite(ffs, b"x", 1, 1, f)
        filefs.fclose(ffs, f)
        filefs.commit(ffs)

    ops["txn_commit"] = time_op(None, txn)

    filefs.umount(ffs)
    filefs.destroy(ffs)
    for p in root.glob("*"):
        p.unlink(missing_ok=True)
    root.rmdir()

    print(
        json.dumps(
            {
                "language": "python",
                "runtime": f"python {sys.version.split()[0]}",
                "ops": ops,
            }
        )
    )


if __name__ == "__main__":
    main()

File: bench/runners/zig/bench.zig
const std = @import("std");
const Io = std.Io;
const filefs = @import("filefs");

const PAYLOAD: usize = 4096;

fn median(samples: []f64) f64 {
    std.mem.sort(f64, samples, {}, comptime std.sort.asc(f64));
    const n = samples.len;
    if (n == 0) return 0;
    if (n % 2 == 0) return (samples[n / 2 - 1] + samples[n / 2]) / 2.0;
    return samples[n / 2];
}

fn nowNs(io: Io) f64 {
    return @floatFromInt(Io.Clock.awake.now(io).nanoseconds);
}

pub fn main(init: std.process.Init) !void {
    const allocator = init.gpa;
    const io = init.io;
    const args = try init.minimal.args.toSlice(init.arena.allocator());
    const iters: usize = if (args.len > 1) try std.fmt.parseInt(usize, args[1], 10) else 40;
    const warmup: usize = if (args.len > 2) try std.fmt.parseInt(usize, args[2], 10) else 2;

    const tmp = "bench-tmp";
    try Io.Dir.cwd().createDirPath(io, tmp);
    defer Io.Dir.cwd().deleteTree(io, tmp) catch {};

    const image = try std.fmt.allocPrint(allocator, "{s}/bench.ffs", .{tmp});
    defer allocator.free(image);
    var counter: usize = 0;

    var payload: [PAYLOAD]u8 = undefined;
    for (&payload, 0..) |*b, i| b.* = @truncate(i);
    var buf: [PAYLOAD]u8 = undefined;

    var out: std.ArrayList(u8) = .empty;
    defer out.deinit(allocator);
    try out.appendSlice(allocator, "{\"language\":\"zig\",\"runtime\":\"zig\",\"ops\":{");
    var first = true;

    const emit = struct {
        fn call(list: *std.ArrayList(u8), gpa: std.mem.Allocator, first_flag: *bool, name: []const u8, ns: f64, n: usize) !void {
            if (!first_flag.*) try list.append(gpa, ',');
            first_flag.* = false;
            try list.print(gpa, "\"{s}\":{{\"ns_per_op\":{d:.3},\"iters\":{d}}}", .{ name, ns, n });
        }
    }.call;

    const Clock = struct {
        var io_handle: Io = undefined;
        fn timeLoop(iters2: usize, warmup2: usize, body: *const fn () anyerror!void) !struct { f64, usize } {
            var i: usize = 0;
            while (i < warmup2) : (i += 1) try body();
            var samples: [512]f64 = undefined;
            const n = @min(iters2, samples.len);
            i = 0;
            while (i < n) : (i += 1) {
                const t0 = nowNs(io_handle);
                try body();
                samples[i] = nowNs(io_handle) - t0;
            }
            return .{ median(samples[0..n]), n };
        }
    };
    Clock.io_handle = io;

    {
        const Ctx = struct {
            var tmp_path: []const u8 = undefined;
            var ctr: *usize = undefined;
            var io_handle: Io = undefined;
            fn body() anyerror!void {
                ctr.* += 1;
                var pbuf: [320]u8 = undefined;
                const p = try std.fmt.bufPrint(&pbuf, "{s}/mkfs{d}.ffs", .{ tmp_path, ctr.* });
                try filefs.mkfs(p);
                Io.Dir.cwd().deleteFile(io_handle, p) catch {};
                var jbuf: [330]u8 = undefined;
                const j = try std.fmt.bufPrint(&jbuf, "{s}-j", .{p});
                Io.Dir.cwd().deleteFile(io_handle, j) catch {};
            }
        };
        Ctx.tmp_path = tmp;
        Ctx.ctr = &counter;
        Ctx.io_handle = io;
        const r = try Clock.timeLoop(iters, warmup, Ctx.body);
        try emit(&out, allocator, &first, "mkfs", r[0], r[1]);
    }

    try filefs.mkfs(image);
    const ffs = try filefs.create(allocator);
    defer ffs.destroy();
    if (!ffs.mount(image)) return error.MountFailed;

    {
        const Ctx = struct {
            var handle: *filefs.FileFS = undefined;
            var image_path: []const u8 = undefined;
            fn body() anyerror!void {
                handle.umount();
                if (!handle.mount(image_path)) return error.MountFailed;
            }
        };
        Ctx.handle = ffs;
        Ctx.image_path = image;
        const r = try Clock.timeLoop(iters, warmup, Ctx.body);
        try emit(&out, allocator, &first, "mount_umount", r[0], r[1]);
    }

    {
        const Ctx = struct {
            var handle: *filefs.FileFS = undefined;
            var ctr: *usize = undefined;
            fn body() anyerror!void {
                ctr.* += 1;
                var nbuf: [64]u8 = undefined;
                const name = try std.fmt.bufPrint(&nbuf, "d{d}", .{ctr.*});
                if (handle.mkdir(name) != 0) return error.MkdirFailed;
            }
        };
        Ctx.handle = ffs;
        Ctx.ctr = &counter;
        const r = try Clock.timeLoop(iters, warmup, Ctx.body);
        try emit(&out, allocator, &first, "mkdir", r[0], r[1]);
    }

    if (ffs.mkdir("cwdbench") != 0) return error.MkdirFailed;
    {
        const Ctx = struct {
            var handle: *filefs.FileFS = undefined;
            fn body() anyerror!void {
                if (!handle.chdir("cwdbench")) return error.ChdirFailed;
                _ = handle.getcwd();
                if (!handle.chdir("/")) return error.ChdirFailed;
            }
        };
        Ctx.handle = ffs;
        const r = try Clock.timeLoop(iters, warmup, Ctx.body);
        try emit(&out, allocator, &first, "chdir_getcwd", r[0], r[1]);
    }

    {
        const Ctx = struct {
            var handle: *filefs.FileFS = undefined;
            var ctr: *usize = undefined;
            fn body() anyerror!void {
                ctr.* += 1;
                var nbuf: [80]u8 = undefined;
                const path = try std.fmt.bufPrint(&nbuf, "o{d}.txt", .{ctr.*});
                const f = handle.fopen(path, "w") orelse return error.OpenFailed;
                handle.fclose(f);
            }
        };
        Ctx.handle = ffs;
        Ctx.ctr = &counter;
        const r = try Clock.timeLoop(iters, warmup, Ctx.body);
        try emit(&out, allocator, &first, "open_write_close", r[0], r[1]);
    }

    {
        const seed = ffs.fopen("seed.bin", "w") orelse return error.OpenFailed;
        if (ffs.fwrite(payload[0..], 1, PAYLOAD, seed) != PAYLOAD) return error.WriteFailed;
        ffs.fclose(seed);
    }

    {
        const Ctx = struct {
            var handle: *filefs.FileFS = undefined;
            var data: *const [PAYLOAD]u8 = undefined;
            fn body() anyerror!void {
                const f = handle.fopen("wbench.bin", "w") orelse return error.OpenFailed;
                if (handle.fwrite(data[0..], 1, PAYLOAD, f) != PAYLOAD) return error.WriteFailed;
                handle.fclose(f);
            }
        };
        Ctx.handle = ffs;
        Ctx.data = &payload;
        const r = try Clock.timeLoop(iters, warmup, Ctx.body);
        try emit(&out, allocator, &first, "write_4kib", r[0], r[1]);
    }

    {
        const Ctx = struct {
            var handle: *filefs.FileFS = undefined;
            var outbuf: *[PAYLOAD]u8 = undefined;
            fn body() anyerror!void {
                const f = handle.fopen("seed.bin", "r") orelse return error.OpenFailed;
                if (handle.fread(outbuf[0..], 1, PAYLOAD, f) != PAYLOAD) return error.ReadFailed;
                handle.fclose(f);
            }
        };
        Ctx.handle = ffs;
        Ctx.outbuf = &buf;
        const r = try Clock.timeLoop(iters, warmup, Ctx.body);
        try emit(&out, allocator, &first, "read_4kib", r[0], r[1]);
    }

    {
        const Ctx = struct {
            var handle: *filefs.FileFS = undefined;
            fn body() anyerror!void {
                const f = handle.fopen("seed.bin", "r") orelse return error.OpenFailed;
                if (!handle.fseek(f, 0, filefs.SeekEnd)) return error.SeekFailed;
                _ = handle.ftell(f);
                handle.rewind(f);
                handle.fclose(f);
            }
        };
        Ctx.handle = ffs;
        const r = try Clock.timeLoop(iters, warmup, Ctx.body);
        try emit(&out, allocator, &first, "seek_tell_rewind", r[0], r[1]);
    }

    {
        const Ctx = struct {
            var handle: *filefs.FileFS = undefined;
            fn body() anyerror!void {
                _ = handle.remove("copy_dst.bin");
                if (handle.copy("seed.bin", "copy_dst.bin") != 0) return error.CopyFailed;
            }
        };
        Ctx.handle = ffs;
        const r = try Clock.timeLoop(iters, warmup, Ctx.body);
        try emit(&out, allocator, &first, "copy_file", r[0], r[1]);
    }

    {
        const Ctx = struct {
            var handle: *filefs.FileFS = undefined;
            var ctr: *usize = undefined;
            fn body() anyerror!void {
                ctr.* += 1;
                var sbuf: [80]u8 = undefined;
                const src = try std.fmt.bufPrint(&sbuf, "r{d}.txt", .{ctr.*});
                ctr.* += 1;
                var dbuf: [80]u8 = undefined;
                const dst = try std.fmt.bufPrint(&dbuf, "s{d}.txt", .{ctr.*});
                const f = handle.fopen(src, "w") orelse return error.OpenFailed;
                handle.fclose(f);
                if (handle.rename(src, dst) != 0) return error.RenameFailed;
                _ = handle.remove(dst);
            }
        };
        Ctx.handle = ffs;
        Ctx.ctr = &counter;
        const r = try Clock.timeLoop(iters, warmup, Ctx.body);
        try emit(&out, allocator, &first, "rename", r[0], r[1]);
    }

    {
        const Ctx = struct {
            var handle: *filefs.FileFS = undefined;
            var ctr: *usize = undefined;
            fn body() anyerror!void {
                ctr.* += 1;
                var nbuf: [80]u8 = undefined;
                const name = try std.fmt.bufPrint(&nbuf, "m{d}.txt", .{ctr.*});
                const f = handle.fopen(name, "w") orelse return error.OpenFailed;
                handle.fclose(f);
                if (handle.remove(name) != 0) return error.RemoveFailed;
            }
        };
        Ctx.handle = ffs;
        Ctx.ctr = &counter;
        const r = try Clock.timeLoop(iters, warmup, Ctx.body);
        try emit(&out, allocator, &first, "remove_file", r[0], r[1]);
    }

    {
        const Ctx = struct {
            var handle: *filefs.FileFS = undefined;
            fn body() anyerror!void {
                const opened = handle.opendir("/") orelse return error.OpenDirFailed;
                const dir = opened.dir;
                while (handle.readdir(dir) != null) {}
                handle.closedir(dir);
            }
        };
        Ctx.handle = ffs;
        const r = try Clock.timeLoop(iters, warmup, Ctx.body);
        try emit(&out, allocator, &first, "readdir", r[0], r[1]);
    }

    {
        const Ctx = struct {
            var handle: *filefs.FileFS = undefined;
            fn body() anyerror!void {
                _ = handle.fileExist("seed.bin");
                _ = handle.dirExist("cwdbench");
            }
        };
        Ctx.handle = ffs;
        const r = try Clock.timeLoop(iters, warmup, Ctx.body);
        try emit(&out, allocator, &first, "exists", r[0], r[1]);
    }

    {
        const Ctx = struct {
            var handle: *filefs.FileFS = undefined;
            var ctr: *usize = undefined;
            fn body() anyerror!void {
                if (!handle.begin()) return error.BeginFailed;
                ctr.* += 1;
                var nbuf: [80]u8 = undefined;
                const name = try std.fmt.bufPrint(&nbuf, "t{d}.txt", .{ctr.*});
                const f = handle.fopen(name, "w") orelse return error.OpenFailed;
                const x: [1]u8 = .{'x'};
                if (handle.fwrite(&x, 1, 1, f) != 1) return error.WriteFailed;
                handle.fclose(f);
                if (!handle.commit()) return error.CommitFailed;
            }
        };
        Ctx.handle = ffs;
        Ctx.ctr = &counter;
        const r = try Clock.timeLoop(iters, warmup, Ctx.body);
        try emit(&out, allocator, &first, "txn_commit", r[0], r[1]);
    }

    ffs.umount();
    try out.appendSlice(allocator, "}}\n");
    var stdout_buffer: [4096]u8 = undefined;
    var stdout_writer = Io.File.stdout().writer(io, &stdout_buffer);
    try stdout_writer.interface.writeAll(out.items);
    try stdout_writer.interface.flush();
}
